### PR TITLE
Output empty `<label/>` instead of omitting it.

### LINKED ISSUE #439: Output empty `<label/>` instead of omitting it.
Currently pyxform requires either a value in the "label" column or in the "hint" column for form controls (ie. questions in the body). This validation rule can remain unchanged.

However, as @tiritea discovered,  a `<label>` element is actually required according to the XForms spec. Therefore, we should include an empty `<label/>` (or `<label></label>`) element in the body output. This is currently not the case.

### Test 1

| type | name | label | hint |
|-----|--------|-----|----|
| text | a        |         | a hint |
| text | b       |A label |        |

```xml
<input ref="/data/a">
    <label/>
    <hint>a hint</hint>
</input>
<input ref="/data/b">
    <label>A label</label>
</input>
```

### PR DESCRIPTION
**Pull request by lindsay-stevens**
- as per #439
- pyxform should always output a label node even if there is no label specified.
- pyxform had a warning for no label but this is removed since there is no requirement for a label.
- added tests to verify that a label is output and no warning raised
- fixed existing tests that either expected a warning or no label

Closes #439

#### Why is this the best possible solution? Were any other approaches considered?

As discussed in the issue ticket, the XForms spec requires a label node. It would not be feasible to change the spec, so this solution brings pyxform output in to line with the spec.

#### What are the regression risks?

It's possible that existing users expect there to be a warning if no label is supplied but a hint is supplied. So the regression risk would be in other apps that use pyxform. In this case the other app could add a check for empty labels.

#### Does this change require updates to documentation? If so, please file an issue [here](https://github.com/XLSForm/xlsform.github.io) and include the link below.

The xlsform.org site says that the survey sheet label is mandatory. However, the update here is specific to pyxform behaviour. It would not be accurate to say that the label is not mandatory, because it is. It's just that for pyxform, in the absence of a label it will emit one in order to comply with the spec.

#### Before submitting this PR, please make sure you have:
- [x] included test cases for core behavior and edge cases in `tests_v1`
- [x] run `nosetests` and verified all tests pass
- [x] run `black pyxform` to format code
- [x] verified that any code or assets from external sources are properly credited in comments

### CODE REVIEW

**Comment by lognaturel** [comment]
From @MartijnR at https://github.com/XLSForm/pyxform/issues/439#issuecomment-867815645:
> If ODK Validate does not get fooled by an empty label (and still requires either a label or hint text content)

The PR description doesn't address this that I can see -- will Validate still throw an error if there's an empty label and no hint? That would be the desired behavior.

**Comment by lognaturel** [comment]
I've come back to try this, and I do confirm that there is still a pyxform error thrown if neither label nor hint is present which is what we want. 👍

I went straight to tests to try to see if that was covered and didn't see it because the forms aren't inline.

File: pyxform/tests_v1/test_guidance_hint.py [hunk 1]
```diff
@@ -73,28 +74,38 @@ def test_multi_language_guidance(self):
 
     def test_guidance_hint_only(self):
         """Test guidance_hint only."""
```
Comment (lognaturel) [inline pyxform/tests_v1/test_guidance_hint.py:76]: This is unexpected. Guidance hints are supplemental so it feels like it doesn't make sense to only have guidance. I guess it's a headache to support that case?

File: pyxform/tests_v1/test_sheet_columns.py [hunk 2]
```diff
@@ -38,6 +41,55 @@ def test_missing_label(self):
             error__contains=["no label or hint"],
         )
 
+    def test_label_node_added_when_hint_given_but_no_label_value(self):
```
Comment (lognaturel) [inline pyxform/tests_v1/test_sheet_columns.py:44]: We've tried as much as possible to move towards tests where the form is inline using `md_to_pyxform_survey`. Is there a strong reason not to do that here? We also have not been introducing new tests on full forms. Was there something that led you to have the test below which I believe is the same but on a real XLSForm?

Comment (lindsay-stevens) [reply pyxform/tests_v1/test_sheet_columns.py:44]: The `PyxformTestCase` machinery seems to simulate from the `workbook_to_json` step onwards. These end to end cases use real XLSX files in case that's too late for the relevant errors or warnings to surface (either now or in future). If preferred I can convert them to the md style test - please let me know.

File: pyxform/tests_v1/test_guidance_hint.py (hunk 1 above)
Comment (lindsay-stevens) [reply pyxform/tests_v1/test_guidance_hint.py:76]: If it's preferred that a guidance hint alone (without regular hint or label) raises an error, please let me know.

**Comment by lindsay-stevens** [comment]
> will Validate still throw an error if there's an empty label and no hint?

For this case there is an existing test: in `test_sheet_columns.InvalidSurveyColumnsTests.test_missing_label`: https://github.com/XLSForm/pyxform/blob/6bb547b409f46f78bb54af398b6da7e869ed418b/pyxform/tests_v1/test_sheet_columns.py#L36

For reference, the two e2e spreadsheets look like this:
hint_with_no_label_column:
hint_with_no_label_column
hint_with_no_label_value:
hint_with_no_label_value

File: pyxform/tests_v1/test_sheet_columns.py (hunk 2 above)
Comment (lindsay-stevens) [reply pyxform/tests_v1/test_sheet_columns.py:44]: One other note - the `xml__contains` option of `PyxformTestCase` takes a sequence of XML snippets and looks for any match in the document. Like issue 448, this test is looking for an exact hierarchy (in case one of the questions doesn't output a label). Once 448 has a solution, this and other output tests would be refactored.

File: pyxform/tests_v1/test_guidance_hint.py (hunk 1 above)
Comment (lognaturel) [reply pyxform/tests_v1/test_guidance_hint.py:76]: Ideally a guidance hint alone does raise an error because they can be configured to be hidden in some clients. If it adds a lot of complexity to have that case, it's not worth it. We know guidance hints are not very common.

Comment (lindsay-stevens) [reply pyxform/tests_v1/test_guidance_hint.py:75]: updated implementation and tests such that an error is thrown for guidance hint only

File: pyxform/tests_v1/test_sheet_columns.py (hunk 2 above)
Comment (lindsay-stevens) [reply pyxform/tests_v1/test_sheet_columns.py:42]: removed e2e tests / files and updated newly added test to markdown style

File: pyxform/survey_element.py [hunk 3]
```diff
@@ -414,19 +420,27 @@ def xml_hint(self):
             )
             return node("hint", hint, toParseString=output_inserted)
 
-    def xml_label_and_hint(self):
+    def xml_label_and_hint(self) -> "List[DetachableElement]":
```
Comment (lognaturel) [inline pyxform/survey_element.py:423]: Spiritually, I'm certainly on board with using type annotations. How are you using them in practice?

File: pyxform/xls2json.py [hunk 4]
```diff
@@ -878,24 +878,6 @@ def workbook_to_json(
             )
             raise PyXFormError(error_message)
 
-        if (
-            constants.LABEL not in row
-            and row.get(constants.MEDIA) is None
-            and question_type not in aliases.label_optional_types
```
Comment (lognaturel) [inline pyxform/xls2json.py:884]: Are there tests for this case and the two below?

File: pyxform/survey_element.py (hunk 3 above)
Comment (lognaturel) [reply pyxform/survey_element.py:423]: From @lindsay-stevens: IntelliJ does verification on an ongoing basis. No effect to runtime.

That seems useful enough to me to add as a tool/convention.

File: pyxform/xls2json.py (hunk 4 above)
Comment (lognaturel) [reply pyxform/xls2json.py:884]: This was for warnings, not for errors. It is related to what @yanokwa was doing at https://github.com/XLSForm/pyxform/pull/542 so we can leave it in here and then look at #542 separately.

Comment (lindsay-stevens) [reply pyxform/xls2json.py:884]: Restored this but in a slightly different location. #542 can still add in the new check condition, and add the tests. Please see commit msg for 3e4a84e

**Comment by lindsay-stevens** [comment]
Hi @lognaturel, resolving this turned out to be not as simple as restoring the check code block. When it was restored, a couple of tests needed to be updated. In particular there is an old-style test with a large "warnings.xls" fixture, which is now refactored into markdown. This should make it clearer when the label warning will and will not be raised. From here, #542 could add in the new condition to exclude warnings for field-list groups, and add the new tests. More context / info in the 3e4a84e commit message.

**Review by lognaturel — APPROVED** [review]
Thank you!
